pygerduty/v2.py

```diff
@@ -0,0 +1,661 @@
+        Container.__init__(self, *args, **kwargs)
+        self.overrides = Overrides(self.pagerduty, self)
+        self.users = ScheduleUsers(self.pagerduty, self)
+        self.entries = Entries(self.pagerduty, self)
+
+
+class ScheduleUser(Container):
+    pass
+
+
+class Team(Container):
+    pass
+
+
+class Restriction(Container):
+    pass
+
+
+class User(Container):
+    def __init__(self, *args, **kwargs):
+        Container.__init__(self, *args, **kwargs)
+        self.notification_rules = NotificationRules(self.pagerduty, self)
+        self.contact_methods = ContactMethods(self.pagerduty, self)
+        self.schedules = Schedules(self.pagerduty, self)
+        self.escalation_policies = EscalationPolicies(self.pagerduty, self)
+        self.log_entries = LogEntries(self.pagerduty, self)
+
+
+class Entry(Container):
+    pass
+
+
+class LogEntry(Container):
+    pass
+
+
+class PagerDuty(object):
+
+    def __init__(self, subdomain, api_token, timeout=10, page_size=25,
+                proxies=None, parse_datetime=False):
+
+        self.api_token = api_token
+        self.subdomain = subdomain
+        self._host = "api.pagerduty.com"
+        self._api_base = "https://{0}/".format(self._host)
+        self.timeout = timeout
+        self.page_size = page_size
+
+        self.json_loader = json.loads
+        if parse_datetime:
+            self.json_loader = _json_loader
+
+        handlers = []
+        if proxies:
+            handlers.append(urllib.request.ProxyHandler(proxies))
+        self.opener = urllib.request.build_opener(*handlers)
+
+        # Collections
+        self.incidents = Incidents(self)
+        self.alerts = Alerts(self)
+        self.schedules = Schedules(self)
+        self.escalation_policies = EscalationPolicies(self)
+        self.users = Users(self)
+        self.services = Services(self)
+        self.maintenance_windows = MaintenanceWindows(self)
+        self.teams = Teams(self)
+        self.log_entries = LogEntries(self)
+
+    @staticmethod
+    def _process_query_params(query_params):
+        new_qp = []
+        for key, value in query_params.items():
+            if isinstance(value, (list, set, tuple)):
+                for elem in value:
+                    new_qp.append(("{0}[]".format(key), elem))
+            else:
+                new_qp.append((key, value))
+
+        return urllib.parse.urlencode(new_qp)
+
+    def request(self, method, path, query_params=None, data=None,
+                extra_headers=None):
+
+        auth = "Token token={0}".format(self.api_token)
+        headers = {
+            "Accept": "application/vnd.pagerduty+json;version=2",
+            "Content-type": "application/json",
+            "Authorization": auth
+        }
+
+        # Requester_id needs to be in the header for v2.
+        if requester_id in data.keys():
+            headers["From"] = data["requester_id"]
+            del data["requester_id"]
```

Comment:
I'm not sure I like the logic in the `request` method as this is mostly a helper that is pretty agnostic to the API itself. In fact this might even be worth pulling out with the other request code.

Callers of `request` would just pass through `.request(..., extra_headers={"From": "meeeee"})`